File: exonum-java-binding/core/rust/src/storage/temporarydb.rs
// Copyright 2018 The Exonum Team
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use exonum::merkledb::{Database, Patch, TemporaryDB};
use jni::{
    objects::{JClass, JObject},
    JNIEnv,
};

use std::panic;

use crate::{
    handle::{self, Handle},
    storage::into_erased_access,
    utils,
};

/// Returns pointer to created `TemporaryDB` object.
#[no_mangle]
pub extern "system" fn Java_com_exonum_binding_core_storage_database_TemporaryDb_nativeCreate(
    env: JNIEnv,
    _: JClass,
) -> Handle {
    let res = panic::catch_unwind(|| Ok(handle::to_handle(TemporaryDB::new())));
    utils::unwrap_exc_or_default(&env, res)
}

/// Destroys underlying `TemporaryDB` object and frees memory.
#[no_mangle]
pub extern "system" fn Java_com_exonum_binding_core_storage_database_TemporaryDb_nativeFree(
    env: JNIEnv,
    _: JClass,
    db_handle: Handle,
) {
    handle::drop_handle::<TemporaryDB>(&env, db_handle);
}

/// Returns pointer to created `Snapshot` object.
#[no_mangle]
pub extern "system" fn Java_com_exonum_binding_core_storage_database_TemporaryDb_nativeCreateSnapshot(
    env: JNIEnv,
    _: JObject,
    db_handle: Handle,
) -> Handle {
    let res = panic::catch_unwind(|| {
        let db = handle::cast_handle::<TemporaryDB>(db_handle);
        let access = unsafe { into_erased_access(db.snapshot()) };
        Ok(handle::to_handle(access))
    });
    utils::unwrap_exc_or_default(&env, res)
}

/// Returns pointer to created `Fork` object.
#[no_mangle]
pub extern "system" fn Java_com_exonum_binding_core_storage_database_TemporaryDb_nativeCreateFork(
    env: JNIEnv,
    _: JObject,
    db_handle: Handle,
) -> Handle {
    let res = panic::catch_unwind(|| {
        let db = handle::cast_handle::<TemporaryDB>(db_handle);
        let access = unsafe { into_erased_access(db.fork()) };
        Ok(handle::to_handle(access))
    });
    utils::unwrap_exc_or_default(&env, res)
}

/// Merges the given patch into the database.
/// The provided `patch_handle` is invalidated after the procedure and the
/// Rust side is responsible for it.
#[no_mangle]
pub extern "system" fn Java_com_exonum_binding_core_storage_database_TemporaryDb_nativeMerge(
    env: JNIEnv,
    _: JObject,
    db_handle: Handle,
    patch_handle: Handle,
) {
    let res = panic::catch_unwind(|| {
        let db = handle::cast_handle::<TemporaryDB>(db_handle);
        let patch = handle::acquire_handle_ownership::<Patch>(patch_handle);
        db.merge(*patch).expect("Unable to merge patch");
        Ok(())
    });
    utils::unwrap_exc_or_default(&env, res)
}

File: exonum-java-binding/core/src/main/java/com/exonum/binding/core/storage/database/TemporaryDb.java
/*
 * Copyright 2019 The Exonum Team
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.exonum.binding.core.storage.database;

import static com.exonum.binding.core.proxy.NativeHandle.INVALID_NATIVE_HANDLE;

import com.exonum.binding.core.proxy.AbstractCloseableNativeProxy;
import com.exonum.binding.core.proxy.Cleaner;
import com.exonum.binding.core.proxy.NativeHandle;
import com.exonum.binding.core.util.LibraryLoader;
import com.google.common.annotations.VisibleForTesting;

/**
 * A MerkleDB which stores its data in the temporary directory for testing purposes.
 * It can create both read-only snapshots and read-write forks. The changes made to
 * database forks can be {@linkplain TemporaryDb#merge(Fork) applied} to the database state.
 *
 * <p>The corresponding database is deleted when TemporaryDb is
 * {@linkplain TemporaryDb#close() closed}.
 *
 * @see com.exonum.binding.core.service.NodeFake
 */
public final class TemporaryDb extends AbstractCloseableNativeProxy implements Database {

  static {
    LibraryLoader.load();
  }

  /**
   * Creates a new empty TemporaryDb.
   */
  public static TemporaryDb newInstance() {
    long nativeHandle = INVALID_NATIVE_HANDLE;
    try {
      nativeHandle = nativeCreate();
      return new TemporaryDb(nativeHandle);
    } catch (Throwable t) {
      if (nativeHandle != INVALID_NATIVE_HANDLE) {
        nativeFree(nativeHandle);
      }
      throw t;
    }
  }

  @VisibleForTesting  // Used in native resource manager tests, must not be exported.
  TemporaryDb(long nativeHandle) {
    super(nativeHandle, true);
  }

  @Override
  public Snapshot createSnapshot(Cleaner cleaner) {
    long snapshotHandle = nativeCreateSnapshot(getNativeHandle());
    return Snapshot.newInstance(snapshotHandle, cleaner);
  }

  @Override
  public Fork createFork(Cleaner cleaner) {
    long forkHandle = nativeCreateFork(getNativeHandle());
    return Fork.newInstance(forkHandle, cleaner);
  }

  /**
   * Applies the changes from the given fork to the database state. TemporaryDb can only
   * merge forks that {@linkplain #createFork(Cleaner) it created itself}.
   *
   * <p>Once this method completes, any indexes created with the fork and the fork itself
   * are closed and cannot be used anymore. Any subsequent operations on these objects will result
   * in {@link IllegalStateException}.
   *
   * <p>If the fork cannot be applied to the database {@link RuntimeException} is
   * thrown and the provided fork is closed.
   *
   * @param fork a fork to get changes from
   */
  public void merge(Fork fork) {
    NativeHandle patchHandle = fork.intoPatch();
    nativeMerge(getNativeHandle(), patchHandle.get());
  }

  @Override
  protected void disposeInternal() {
    nativeFree(getNativeHandle());
  }

  private static native long nativeCreate();

  private native long nativeCreateSnapshot(long dbNativeHandle);

  private native long nativeCreateFork(long dbNativeHandle);

  private native void nativeMerge(long dbNativeHandle, long patchNativeHandle);

  private static native void nativeFree(long dbNativeHandle);
}
